fix: Inproper signal handling

mowirth opened the pull request:
<!-- Prepend '[WIP]' to the title if this PR is still a work-in-progress. Remove it when it is ready for review! -->

## Description

Fixes #4839.

Replaces the current implementation with a helper implementation to avoid os.

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [x] I authored this code entirely myself
- [x] I am submitting code based on open source software (e.g. MIT, MPL-2.0, Apache)]
- [x] I am adding or upgrading a dependency or adapted code and confirm it has a compatible open source license
- [x] Update the docs.
- [ ] Run the relevant tests successfully, including pre-commit checks.
- [ ] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

- Fixed regression regarding the signal handling for subprocesses

### Migration Guide

No incompatibilities


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added graceful cancellation for spawned processes, enabling interruption via an interrupt signal.

* **Tests**
  * Improved test behavior and diagnostics for interrupted/killed processes on Unix and Windows.

* **Chores**
  * Minor import and formatting cleanup across the codebase.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

mowirth commented:
I have some issues running the test locally as I have not installed all dependencies, so I'll be happy if you could give them a run and check if something is not in order.

mowirth commented:
@sylr thanks for the help, I integrated it in in my branch.

Regarding the lint errors related to the CommandContext, I don't think there is a way around this - as expained above, CommandContext will always send a SIGKILL to the underlying process.

jordanst3wart commented on internal/os/exec/cmd.go:
```diff
@@ -128,3 +129,21 @@ func (cmd *Cmd) SendSignal(sig os.Signal) {
+		if cmd.Process != nil {
+			// Send the provided signal to the process group
+			_ = syscall.Kill(-cmd.Process.Pid, signal.InterruptSignal)
+		}
```
This doesn't respect the context's cancel function

mowirth commented:
Apologies for the delay, I've been tied up the last days.

I refactored the fix again and switched to overwriting the cancel function - the strict-lint runs now and the tests look good to.

Please feel free to have another look

mowirth commented:
Please take another look and cross check the Codebrabbit stuff regarding the signal handling and the error wrapping.
